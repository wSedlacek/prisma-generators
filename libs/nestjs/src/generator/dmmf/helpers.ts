// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
