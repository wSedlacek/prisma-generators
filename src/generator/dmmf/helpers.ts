// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
