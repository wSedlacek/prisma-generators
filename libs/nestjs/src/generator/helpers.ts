import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
