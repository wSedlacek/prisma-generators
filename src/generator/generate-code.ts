import { DMMF as PrismaDMMF } from "@prisma/client/runtime/dmmf-types";
import { Project } from "ts-morph";
import path from "path";

import { noop } from "./helpers";
import generateEnumFromDef from "./enum";
import generateObjectTypeClassFromModel from "./model-type-class";
import generateRelationsResolverClassesFromModel from "./resolvers/relations";
import {
  generateOutputTypeClassFromType,
  generateInputTypeClassFromType,
} from "./type-class";
import generateCrudResolverClassFromMapping from "./resolvers/full-crud";
import {
  resolversFolderName,
  relationsResolversFolderName,
  crudResolversFolderName,
  inputsFolderName,
  outputsFolderName,
  enumsFolderName,
  modelsFolderName,
  argsFolderName,
} from "./config";
import {
  generateResolversBarrelFile,
  generateInputsBarrelFile,
  generateOutputsBarrelFile,
  generateIndexFile,
  generateModelsBarrelFile,
  generateEnumsBarrelFile,
  generateArgsBarrelFile,
} from "./imports";
import saveSourceFile from "../utils/saveSourceFile";
import { GenerateCodeOptions } from "./options";
import { DmmfDocument } from "./dmmf/dmmf-document";

export default async function generateCode(
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop,
) {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map(model => model.name);

  log("Transforming dmmfDocument...");
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log("Generating enums...");
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    enumDef => enumDef.name,
  );
  await Promise.all(
    dmmfDocument.datamodel.enums.map(enumDef =>
      generateEnumFromDef(project, baseDirPath, enumDef),
    ),
  );
  await Promise.all(
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter(enumDef => !datamodelEnumNames.includes(enumDef.name))
      .map(enumDef => generateEnumFromDef(project, baseDirPath, enumDef)),
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map(it => it.name),
      ...dmmfDocument.datamodel.enums.map(it => it.name),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log("Generating models...");
  await Promise.all(
    dmmfDocument.datamodel.models.map(model =>
      generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        dmmfDocument,
        options,
      ),
    ),
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, "index.ts"),
    undefined,
    { overwrite: true },
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map(it => it.typeName),
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log("Generating output types...");
  const rootTypes = dmmfDocument.schema.outputTypes.filter(type =>
    ["Query", "Mutation"].includes(type.name),
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    type => !modelNames.includes(type.name) && !rootTypes.includes(type),
  );
  await Promise.all(
    outputTypesToGenerate.map(type =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options,
      ),
    ),
  );
  const argsTypesNames = outputTypesToGenerate
    .map(it => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .map(it => it.argsTypeName)
    .filter(Boolean) as string[];
  if (argsTypesNames.length > 0) {
    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateArgsBarrelFile(outputsArgsBarrelExportSourceFile, argsTypesNames);
    await saveSourceFile(outputsArgsBarrelExportSourceFile);
  }

  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map(it => it.typeName),
    argsTypesNames.length > 0,
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  log("Generating input types...");
  await Promise.all(
    dmmfDocument.schema.inputTypes.map(type =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options,
      ),
    ),
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map(it => it.typeName),
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log("Generating relation resolvers...");
  const relationResolversData = await Promise.all(
    dmmfDocument.datamodel.models
      .filter(model => model.fields.some(field => field.relationName))
      .map(model => {
        const outputType = dmmfDocument.schema.outputTypes.find(
          type => type.name === model.name,
        )!;
        const mapping = dmmfDocument.mappings.find(
          it => it.model === model.name,
        )!;
        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          dmmfDocument,
        );
      }),
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      "relations",
      relationResolversBarrelExportSourceFile,
      relationResolversData,
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log("Generating crud resolvers...");
  const crudResolversData = await Promise.all(
    dmmfDocument.mappings.map(mapping => {
      const model = dmmfDocument.datamodel.models.find(
        model => model.name === mapping.model,
      )!;
      return generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        rootTypes,
        modelNames,
        options,
        dmmfDocument,
      );
    }),
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      "index.ts",
    ),
    undefined,
    { overwrite: true },
  );
  generateResolversBarrelFile(
    "crud",
    crudResolversBarrelExportSourceFile,
    crudResolversData,
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log("Generating index file");
  const indexSourceFile = project.createSourceFile(
    baseDirPath + "/index.ts",
    undefined,
    { overwrite: true },
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
}
