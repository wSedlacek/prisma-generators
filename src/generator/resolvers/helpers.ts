import { getFieldTSType, getGraphQLType } from '../helpers';
import { DmmfDocument } from '../dmmf/dmmf-document';
import { DMMF } from '../dmmf/types';

export const generateCrudResolverClassMethodDeclaration = (
  action: DMMF.Action,
  typeName: string,
  method: DMMF.SchemaField,
  argsTypeName: string | undefined,
  collectionName: string,
  dmmfDocument: DmmfDocument,
  mapping: DMMF.Mapping
) => {
  const returnTSType = getFieldTSType(
    method.outputType,
    dmmfDocument,
    false,
    mapping.model,
    typeName
  );

  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${returnTSType}>`,
    decorators: [
      {
        name: `${action.operation}`,
        arguments: [
          `() => ${getGraphQLType(
            method.outputType,
            dmmfDocument,
            mapping.model,
            typeName
          )}`,
          `{
            ${[
              `nullable: ${!method.outputType.isRequired}`,
              `description: undefined`,
            ].join(',\n')}
          }`,
        ],
      },
    ],
    parameters: [
      {
        name: 'ctx',
        // TODO: import custom `ContextType`
        type: 'any',
        decorators: [{ name: 'Context', arguments: [] }],
      },
      ...(action.kind === 'aggregate'
        ? [
            {
              name: 'info',
              type: 'GraphQLResolveInfo',
              decorators: [{ name: 'Info', arguments: [] }],
            },
          ]
        : []),
      ...(!argsTypeName
        ? []
        : [
            {
              name: 'args',
              type: argsTypeName,
              decorators: [{ name: 'Args', arguments: [] }],
            },
          ]),
    ],
    statements:
      action.kind === 'aggregate'
        ? [
            `const transformFields = (fields: Record<string, any>): Record<string, any> => {
              return Object.fromEntries(
                Object.entries(fields)
                  .filter(([key, value]) => !key.startsWith("_"))
                  .map<[string, any]>(([key, value]) => {
                    if (Object.keys(value).length === 0) {
                      return [key, true];
                    }
                    return [key, transformFields(value)];
                  }),
              );
            }`,
            `return ctx.prisma.${collectionName}.${action.kind}({
              ...args,
              ...transformFields(graphqlFields(info as any)),
            });`,
          ]
        : [
            `return plainToClass(${returnTSType
              .replace(/\[|\]/g, '')
              .replace(
                '| undefined',
                ''
              )}, await ctx.prisma.${collectionName}.${action.kind}(${
              argsTypeName ? 'args' : ''
            }) as ${getGraphQLType(
              method.outputType,
              dmmfDocument,
              mapping.model,
              typeName
            )});`,
          ],
  };
};
