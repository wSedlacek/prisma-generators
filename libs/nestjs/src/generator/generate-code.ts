import { DMMF as PrismaDMMF } from '@prisma/client/runtime/dmmf-types';
import path from 'path';
import { Project } from 'ts-morph';

import { isDefined, saveSourceFile } from '../utils';
import { generateArgsTypeClassFromArgs } from './args-class';
import {
  argsFolderName,
  crudResolversFolderName,
  enumsFolderName,
  inputsFolderName,
  modelsFolderName,
  outputsFolderName,
  relationsResolversFolderName,
  resolversFolderName,
} from './config';
import { DmmfDocument } from './dmmf/dmmf-document';
import { DMMF } from './dmmf/types';
import { generateEnumFromDef } from './enum';
import { noop } from './helpers';
import {
  generateArgsBarrelFile,
  generateEnumsBarrelFile,
  generateIndexFile,
  generateInputsBarrelFile,
  generateModelsBarrelFile,
  generateOutputsBarrelFile,
  generateResolversBarrelFile,
} from './imports';
import { generateObjectTypeClassFromModel } from './model-type-class';
import { GenerateCodeOptions } from './options';
import { generateCrudResolverClassFromMapping } from './resolvers/full-crud';
import { generateRelationsResolverClassesFromModel } from './resolvers/relations';
import {
  generateInputTypeClassFromType,
  generateOutputTypeClassFromType,
} from './type-class';

export const generateCode = async (
  dmmf: PrismaDMMF.Document,
  options: GenerateCodeOptions,
  log: (msg: string) => void = noop
) => {
  const baseDirPath = options.outputDirPath;
  const project = new Project();
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);
  const modelNames = dmmf.datamodel.models.map((model) => model.name);

  log('Transforming dmmfDocument...');
  const dmmfDocument = new DmmfDocument(dmmf, options);

  log('Generating enums...');
  const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
    (enumDef) => enumDef.typeName
  );
  await Promise.all(
    dmmfDocument.datamodel.enums.map((enumDef) =>
      generateEnumFromDef(project, baseDirPath, enumDef)
    )
  );
  await Promise.all(
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter((enumDef) => !datamodelEnumNames.includes(enumDef.typeName))
      .map((enumDef) => generateEnumFromDef(project, baseDirPath, enumDef))
  );
  const emittedEnumNames = [
    ...new Set([
      ...dmmfDocument.schema.enums.map((it) => it.typeName),
      ...dmmfDocument.datamodel.enums.map((it) => it.typeName),
    ]),
  ];
  const enumsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, enumsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  await saveSourceFile(enumsBarrelExportSourceFile);

  log('Generating models...');
  await Promise.all(
    dmmfDocument.datamodel.models.map((model) =>
      generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        dmmfDocument,
        options
      )
    )
  );
  const modelsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(baseDirPath, modelsFolderName, 'index.ts'),
    undefined,
    { overwrite: true }
  );
  generateModelsBarrelFile(
    modelsBarrelExportSourceFile,
    dmmfDocument.datamodel.models.map((it) => it.typeName)
  );
  await saveSourceFile(modelsBarrelExportSourceFile);

  log('Generating output types...');
  const rootTypes = dmmfDocument.schema.outputTypes.filter((type) =>
    ['Query', 'Mutation'].includes(type.name)
  );
  const outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
    // skip generating models and root resolvers
    (type) => !modelNames.includes(type.name) && !rootTypes.includes(type)
  );
  const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
    .map((it) => it.fields)
    .reduce((a, b) => a.concat(b), [])
    .filter((it): it is Required<DMMF.OutputSchemaField> => !!it.argsTypeName);
  await Promise.all(
    outputTypesToGenerate.map((type) =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const outputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      outputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateOutputsBarrelFile(
    outputsBarrelExportSourceFile,
    outputTypesToGenerate.map((it) => it.typeName),
    outputTypesFieldsArgsToGenerate.length > 0
  );
  await saveSourceFile(outputsBarrelExportSourceFile);

  if (outputTypesFieldsArgsToGenerate.length > 0) {
    log('Generating output types args...');
    await Promise.all(
      outputTypesFieldsArgsToGenerate.map(async (field) => {
        await generateArgsTypeClassFromArgs(
          project,
          path.resolve(resolversDirPath, outputsFolderName),
          field.args,
          field.argsTypeName,
          dmmfDocument,
          options,
          2
        );
      })
    );
    const outputsArgsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        argsFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateArgsBarrelFile(
      outputsArgsBarrelExportSourceFile,
      outputTypesFieldsArgsToGenerate
        .map((it) => it.argsTypeName)
        .filter(isDefined)
    );
    await saveSourceFile(outputsArgsBarrelExportSourceFile);
  }

  log('Generating input types...');
  await Promise.all(
    dmmfDocument.schema.inputTypes.map((type) =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options
      )
    )
  );
  const inputsBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      inputsFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateInputsBarrelFile(
    inputsBarrelExportSourceFile,
    dmmfDocument.schema.inputTypes.map((it) => it.typeName)
  );
  await saveSourceFile(inputsBarrelExportSourceFile);

  log('Generating relation resolvers...');
  const relationResolversData = await Promise.all(
    dmmfDocument.datamodel.models
      .filter((model) => model.fields.some((field) => field.relationName))
      .map(async (model) => {
        const outputType = dmmfDocument.schema.outputTypes.find(
          (type) => type.name === model.name
        );
        const mapping = dmmfDocument.mappings.find(
          (it) => it.model === model.name
        );

        if (!outputType || !mapping) {
          throw new Error('Could not determine resolver data');
        }

        return generateRelationsResolverClassesFromModel(
          project,
          baseDirPath,
          model,
          mapping,
          outputType,
          dmmfDocument,
          options
        );
      })
  );
  if (relationResolversData.length > 0) {
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        'index.ts'
      ),
      undefined,
      { overwrite: true }
    );
    generateResolversBarrelFile(
      'relations',
      relationResolversBarrelExportSourceFile,
      relationResolversData
    );
    await saveSourceFile(relationResolversBarrelExportSourceFile);
  }

  log('Generating crud resolvers...');
  const crudResolversData = await Promise.all(
    dmmfDocument.mappings.map(async (mapping) => {
      const model = dmmfDocument.datamodel.models.find(
        (it) => it.name === mapping.model
      );

      if (!model) {
        throw new Error('Could not determine resolver data');
      }

      return generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        modelNames,
        options,
        dmmfDocument
      );
    })
  );
  const crudResolversBarrelExportSourceFile = project.createSourceFile(
    path.resolve(
      baseDirPath,
      resolversFolderName,
      crudResolversFolderName,
      'index.ts'
    ),
    undefined,
    { overwrite: true }
  );
  generateResolversBarrelFile(
    'crud',
    crudResolversBarrelExportSourceFile,
    crudResolversData
  );
  await saveSourceFile(crudResolversBarrelExportSourceFile);

  log('Generating index file');
  const indexSourceFile = project.createSourceFile(
    `${baseDirPath}/index.ts`,
    undefined,
    { overwrite: true }
  );
  generateIndexFile(indexSourceFile, relationResolversData.length > 0);
  await saveSourceFile(indexSourceFile);
};
