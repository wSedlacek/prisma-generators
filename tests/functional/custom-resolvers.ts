import "reflect-metadata";
import { promises as fs } from "fs";
import { buildSchema, Query, Resolver, Args, Ctx } from "type-graphql";
import { graphql } from "graphql";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";

describe("custom resolvers execution", () => {
  let outputDirPath: string;

  beforeAll(async () => {
    outputDirPath = generateArtifactsDirPath("functional-custom-resolvers");
    await fs.mkdir(outputDirPath, { recursive: true });
    const prismaSchema = /* prisma */ `
      enum Color {
        RED
        GREEN
        BLUE
      }

      model Post {
        uuid     String  @id @default(cuid())
        content  String
        color    Color
      }
    `;
    await generateCodeFromSchema(prismaSchema, { outputDirPath });
  });

  it("should be possible to use generated inputs, args and types to build own resolvers", async () => {
    const { Post, FindManyPostArgs } = require(outputDirPath);
    @Resolver()
    class CustomResolver {
      @Query(_returns => [Post])
      async customFindManyPost(
        @Args(_type => FindManyPostArgs) args: any,
        @Ctx() { prisma }: any,
      ) {
        return await prisma.post.findMany(args);
      }
    }
    const document = /* graphql */ `
      query {
        customFindManyPost(
          take: 1
          skip: 1
          where: {
            content: { startsWith: "Test" }
          }
          orderBy: {
            color: desc
          }
        ) {
          uuid
          color
        }
      }
    `;
    const prismaMock = {
      post: {
        findMany: jest.fn().mockResolvedValue([
          {
            uuid: "b0c0d78e-4dff-4cdd-ba23-9b417dc684e2",
            color: "RED",
          },
          {
            uuid: "72c8a188-3d46-45b3-b23f-7d420aa282f1",
            color: "BLUE",
          },
        ]),
      },
    };

    const graphQLSchema = await buildSchema({
      resolvers: [CustomResolver],
      validate: false,
      emitSchemaFile: outputDirPath + "/schema.graphql",
    });
    const graphQLSchemaSDL = await fs.readFile(
      outputDirPath + "/schema.graphql",
      { encoding: "utf8" },
    );

    const { data, errors } = await graphql(graphQLSchema, document, null, {
      prisma: prismaMock,
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("custom posts resolver mocked response");
    expect(prismaMock.post.findMany.mock.calls).toMatchSnapshot(
      "findManyPost call args",
    );
    expect(graphQLSchemaSDL).toMatchSnapshot("graphQLSchemaSDL");
  });
});
