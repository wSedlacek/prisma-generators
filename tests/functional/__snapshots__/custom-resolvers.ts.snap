// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom resolvers execution should be possible to use generated inputs, args and types to build own resolvers: custom posts resolver mocked response 1`] = `
Object {
  "customFindManyPost": Array [
    Object {
      "color": "RED",
      "uuid": "b0c0d78e-4dff-4cdd-ba23-9b417dc684e2",
    },
    Object {
      "color": "BLUE",
      "uuid": "72c8a188-3d46-45b3-b23f-7d420aa282f1",
    },
  ],
}
`;

exports[`custom resolvers execution should be possible to use generated inputs, args and types to build own resolvers: findManyPost call args 1`] = `
Array [
  Array [
    FindManyPostArgs {
      "orderBy": PostOrderByInput {
        "color": "desc",
      },
      "skip": 1,
      "take": 1,
      "where": PostWhereInput {
        "content": StringFilter {
          "startsWith": "Test",
        },
      },
    },
  ],
]
`;

exports[`custom resolvers execution should be possible to use generated inputs, args and types to build own resolvers: graphQLSchemaSDL 1`] = `
"# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

enum Color {
  BLUE
  GREEN
  RED
}

input ColorFilter {
  equals: Color
  in: [Color!]
  not: Color
  notIn: [Color!]
}

enum OrderByArg {
  asc
  desc
}

type Post {
  color: Color!
  content: String!
  uuid: String!
}

input PostOrderByInput {
  color: OrderByArg
  content: OrderByArg
  uuid: OrderByArg
}

input PostWhereInput {
  AND: [PostWhereInput!]
  color: ColorFilter
  content: StringFilter
  NOT: [PostWhereInput!]
  OR: [PostWhereInput!]
  uuid: StringFilter
}

input PostWhereUniqueInput {
  uuid: String
}

type Query {
  customFindManyPost(cursor: PostWhereUniqueInput, orderBy: PostOrderByInput, skip: Int, take: Int, where: PostWhereInput): [Post!]!
}

input StringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: String
  notIn: [String!]
  startsWith: String
}
"
`;
